### ISSUE TITLE
Consider: Removing StatusOr from Client::Read and Client::ExecuteSql

### ISSUE THREAD

**Issue by devjgm** [googleapis/google-cloud-cpp-spanner#628]
Consider the following code that a caller would need to write today:

```cc
Client c = ...;
StatusOr<ResultSet> results = c.Read(...);
if (!results) {
  // handle error  <-- 1
}
for (auto row : results.Rows<T1, T2>()) {
  if (!row) {
    // handle error  <-- 2
  }
  ...
}
```

There are two `StatusOr` objects that the caller needs to deal with in this flow (indicated in the code above). It would be nice to remove the first `StatusOr` that wraps the `ResultSet` returned from `Client::Read` and `Client::ExecuteSql`, so that the only `StatusOr` that the caller needs to deal with is the one inside the loop.

The advantage of this is that it simplifies the caller's code. One disadvantage is that it makes these functions different from the other functions on `Client`, which all return `StatusOr<T>`.

Questions:
1. Is it possible to remove this outter StatusOr?
2. Do we want to do this?

**Comment by coryan** [comment]
I think the problem is:

```C++
ResultSet result = client->ExecuteSql("INSERT INTO Table ...");
// How do you check the status?  Cannot call `Rows<Ts>` for any `Ts`
```

**Comment by devjgm** [comment]
Yep, that's related to another issue I was filing https://github.com/googleapis/google-cloud-cpp-spanner/issues/629

**Comment by mr-salty** [comment]
Greg, Scott, and I spent some time discussing this and concluded we should do it, although #629 should happen first because the DML calls do still need to return Status as Carlos noted above.

We had some concerns about possible bad interactions with deferred Transaction creation, but reading the code for `Visit`, etc. it does not appear to be an issue. Specifically, if we make multiple calls with the same `Transaction`, then the ones that succeed are still guaranteed to use the same transaction from Spanner's perspective, i.e. given.

```
Transaction ro_tx = ...;
auto r1 = client.Read(ro_tx, ...);
auto r2 = client.Read(ro_tx, ...);
```

If the first `Read()` fails then `ro_tx` will not have a transaction ID when the second `Read()` occurs, so that call will also attempt to begin the transaction.  Locking inside `Visit` ensures the transaction creation is serialized even if the `Read` calls happen in different threads.

**Comment by devjgm** [comment]
@scotthart I assigned this to you only because I *think* you're already doing it. If I'm wrong and you're not already working on this (and almost done with it), feel free to unassign this.
